### ISSUE TITLE
Spawning error

### ISSUE THREAD

**Issue by lovelydinosaur** [Kludex/uvicorn#315]
Attempting to use reload, can result in this error:

```python
Traceback (most recent call last):
  File "example.py", line 53, in <module>
    uvicorn.run(app.__call__, reload=True)
  File "/Users/tomchristie/GitHub/encode/typesystem/venv/lib/python3.6/site-packages/uvicorn/main.py", line 255, in run
    supervisor.run(server.run, sockets=[socket])
  File "/Users/tomchristie/GitHub/encode/typesystem/venv/lib/python3.6/site-packages/uvicorn/supervisors/statreload.py", line 34, in run
    process.start()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/process.py", line 105, in start
    self._popen = self._Popen(self)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/popen_fork.py", line 20, in __init__
    self._launch(process_obj)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'request_response.<locals>.app'
```

Will look into this first thing tomorrow.

**Comment by florimondmanca** [comment]
+1 to this with the following Starlette app:

```python
from starlette.applications import Starlette
from starlette.responses import JSONResponse
import uvicorn

app = Starlette()


@app.route("/")
async def homepage(request):
    return JSONResponse({"hello": "world"})


if __name__ == "__main__":
    uvicorn.run(app, port=8000, reload=True)
```

```bash
$ python app.py
INFO: Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO: Started reloader process [30701]
Traceback (most recent call last):
  File "app.py", line 14, in <module>
    uvicorn.run(app, port=8000, reload=True)
  File "/Users/Florimond/.pyenv/versions/3.7.2/lib/python3.7/site-packages/uvicorn/main.py", line 255, in run
    supervisor.run(server.run, sockets=[socket])
  File "/Users/Florimond/.pyenv/versions/3.7.2/lib/python3.7/site-packages/uvicorn/supervisors/statreload.py", line 34, in run
    process.start()
  File "/Users/Florimond/.pyenv/versions/3.7.2/lib/python3.7/multiprocessing/process.py", line 112, in start
    self._popen = self._Popen(self)
  File "/Users/Florimond/.pyenv/versions/3.7.2/lib/python3.7/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/Florimond/.pyenv/versions/3.7.2/lib/python3.7/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/Florimond/.pyenv/versions/3.7.2/lib/python3.7/multiprocessing/popen_fork.py", line 20, in __init__
    self._launch(process_obj)
  File "/Users/Florimond/.pyenv/versions/3.7.2/lib/python3.7/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/Florimond/.pyenv/versions/3.7.2/lib/python3.7/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'request_response.<locals>.app'
```

**Comment by lovelydinosaur** [comment]
Okay. That's a total PITA - works great in the test cases. And will work in *some* examples, but not others. Reason is that it turns out the way Python's multiprocessing calls work is by serializing/deserializing the arguments to the call using pickle. Eveidently in some cases there'll end up being something somewhere referenced by the application that's not picklable.

You won't see these errors running from the command line, because the application instance isn't passed into the call - instead an import string is passed. eg `example:app`.

**Comment by lovelydinosaur** [comment]
The simplest fix will be to switch back from this... (0.5)

```python
spawn = multiprocessing.get_context("spawn")
process = spawn.Process(target=target, args=args, kwargs=kwargs)
```

To this...

```python
process = multiprocessing.Process(target=target, args=args, kwargs=kwargs)
```

Which'll fix the error, but will mean that we're back to reload not working when used programmatically.

(Unless you pass an import string, rather than the instance itself)

**Comment by lovelydinosaur** [comment]
Best we can do is this: https://github.com/encode/uvicorn/pull/317

This disables auto-reload except in cases where the app is passed as an import string.

**Comment by lovelydinosaur** [comment]
Resolved in 0.5.1
